tests/functional/history/test_db.py

```diff
@@ -0,0 +1,311 @@
+                }, {
+                    'command_id': 'command',
+                    'event_type': 'HTTP_REQUEST',
+                    'payload': i,
+                    'source': 'TEST',
+                    'timestamp': 1234
+                }, {
+                    'command_id': 'command',
+                    'event_type': 'HTTP_RESPONSE',
+                    'payload': i,
+                    'source': 'TEST',
+                    'timestamp': 1234
+                }, {
+                    'command_id': 'command',
+                    'event_type': 'PARSED_RESPONSE',
+                    'payload': i,
+                    'source': 'TEST',
+                    'timestamp': 1234
+                }
+            ])
+        for t in self.threads:
+            t.close()
+        thread_id_to_request_id = {}
+        cursor = self.connection.cursor()
+        cursor.execute('select request_id, payload from records')
+        records = 0
+        for record in cursor:
+            records += 1
+            request_id = record['request_id']
+            thread_id = record['payload']
+            if thread_id not in thread_id_to_request_id:
+                thread_id_to_request_id[thread_id] = request_id
+            else:
+                prior_request_id = thread_id_to_request_id[thread_id]
+                self.assertEqual(request_id, prior_request_id)
+        self.assertEqual(records, 4 * thread_count)
+
+
+@unittest.skipIf(sqlite3 is None,
+                 "sqlite3 not supported in this python")
+class TestDatabaseRecordWriter(BaseDatabaseTest):
+    def test_does_create_database(self):
+        self.assertTrue(os.path.isfile(self.db_filename))
+
+    def test_does_create_table(self):
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM sqlite_master WHERE "
+                       "type='table' AND name='records';")
+        result = cursor.fetchone()
+        self.assertEqual(result[0], 1)
+
+    def test_can_write_record(self):
+        writer = DatabaseRecordWriter(connection=self.connection)
+        known_record_fields = {
+            'command_id': 'command',
+            'source': 'TEST',
+            'event_type': 'foo',
+            'payload': {"foo": "bar"},
+            'timestamp': 1234
+        }
+        writer.write_record(known_record_fields)
+
+        cursor = self._get_cursor()
+        cursor.execute("SELECT COUNT(*) FROM records;")
+        num_records = cursor.fetchone()
+        self.assertEqual(num_records[0], 1)
+
+        cursor.execute("SELECT * FROM records;")
+        record = dict(cursor.fetchone())
+        for col_name, row_value in known_record_fields.items():
+            # Normally our reader would take care of parsing the JSON from
+            # the payload.
+            if col_name == 'payload':
+                record[col_name] = json.loads(record[col_name])
+                self.assertEqual(record[col_name], row_value)
+
+        self.assertIn('id', record.keys())
```

Comment:
Usually you do not have to use `keys()`. `in` for a dict means to look in the keys for it already.